pkg/status/status.go

```diff
@@ -191,12 +269,99 @@ func (s *status) setStatusCondition(condition configv1.ClusterStatusConditionTyp
+
+		// Check if the clusterOperator version has changed and log the attempt to upgrade if it has
+		previousVersion := operatorhelpers.FindOperandVersion(previousStatus.Versions, "operator")
+		currentVersion := operatorhelpers.FindOperandVersion(s.clusterOperator.Status.Versions, "operator")
+		if currentVersion != nil {
+			if previousVersion == nil {
+				log.Infof("[status] Attempting to set clusterOperator to version %s", currentVersion.Version)
+			} else if previousVersion.Version != currentVersion.Version {
+				log.Infof("[status] Attempting to upgrade clusterOperator version from %s to %s", previousVersion.Version, currentVersion.Version)
+			}
+		}
+
+		_, err := s.configClient.ClusterOperators().UpdateStatus(s.clusterOperator)
+		if err != nil {
+			return fmt.Errorf("Error %v updating ClusterOperator", err)
+		}
+		// Log status change
+		log.Infof(fmt.Sprintf("[status] Set ClusterOperator condition: %s message: %s", updatedCondition, message))
+	}
+	return err
+}
+
+// syncChannelReceiver will listen on the sync channel and update the status'
+// syncsRatio filed until the stopCh is closed.
+func (s *status) syncChannelReceiver() {
+	log.Info("[status] Starting sync consumer")
+	for {
+		select {
+		case <-s.stopCh:
+			return
+		case err := <-s.syncCh:
+			if err == nil {
+				s.syncRatio.ReportSyncEvent()
+			} else {
+				s.syncRatio.ReportFailedSync()
+			}
+			failedSyncs, syncs := s.syncRatio.GetSyncs()
+			log.Debugf("[status] Faild Syncs / Total Syncs : %d/%d", failedSyncs, syncs)
+		}
+	}
+}
+
+// monitorClusterStatus updates the clusteroperator's status based on
+// the number of successful syncs / total syncs
+func (s *status) monitorClusterStatus() {
+	// Signal to the main channel that we have stopped reporting status.
+	defer func() {
+		close(s.monitorDoneCh)
+	}()
+	for {
+		select {
+		case <-s.stopCh:
+			// If the stopCh is closed, the operator will exit and CO should
+			// be set to failing.
+			s.setFailing("Operator exited")
+			return
+		// Attempt to update the status every 20 seconds
+		case <-time.After(20 * time.Second):
+			var statusErr error
+			// create the cluster operator in the porgressing state if it does not exist
+			// or if it is the first report
+			if s.clusterOperator == nil {
+				statusErr = s.setProgressing(fmt.Sprintf("Progressing towards %s", s.version))
+			} else {
+				// wait until the operator has reconciled at least one sync
+				_, syncEvents := s.syncRatio.GetSyncs()
+				if syncEvents >= minSyncsBeforeReporting {
+					// update the status with the appropriate state
+					isSucceeding, err := s.syncRatio.IsSucceeding()
+					if err == nil {
+						if isSucceeding {
+							statusErr = s.setAvailable(fmt.Sprintf("%s is available", s.version))
+						} else {
+							statusErr = s.setFailing("The operator is not meeting the expected sync ratio.")
```

Comment:
It will help if you add useful data in the message.

Reply:
I will update the message to provide better information about the sync ratio.

Planned Update:
current sync ratio (%g) does not meet the expected success threshold (%g)